README
Multiscale Entropic Network Generator 2 (MUSKETEER2)

Copyright (c) 2011-2018 by Alexander Gutfraind and Ilya Safro. 
All rights reserved.

Use and redistribution of this file is governed by the license terms in
the LICENSE file found in the project's top-level directory.

Setup and Installation
----------------------
Make sure your system has 
1. Python 2.6 or 2.7
2. NetworkX library (version >= 1.10)
3. numpy library (version >=1.5)

You can check this by running 

python

From Python:

import sys, numpy, networkx
sys.version
numpy.__version__
networkx.__version__

Also,
graphviz is recommended for visualization.



Usage example
-------------
python musketeer.py -f data-samples/arenas_email.edges -t edgelist -p "{'node_growth_rate':[0.005, 0.001], 'edge_edit_rate':[0.05, 0.04, 0.03], 'node_edit_rate':[0.07, 0.06, 0.05]}" -o output/test.dot

explanation:
1. Loads data-social/arenas_email.edges
2. Increases the number of nodes by 0.9% at level 1 deep, and by 0.1% at level 0 (finest level)
3. Edits 3% of the edges and 5% of the nodes at level 2 deep, 
   4% of edges at level 1 deep and 6% of the nodes,
   5% of edges at level 0 deep and 7% of the nodes
4. The graph is outputed to output/test.dot (DOT format used by graphviz)

#note that step 2 changes the network statistics, and is often skipped in practice

#a direct call from python script:
import algorithms
new_G = algorithms.generate_graph(G, params)



Special usage modes
--------------------
node_growth_rate:
This parameter accepts values in (-1,infinity) at every level.  
When set to values 1 or greater it will establish number_of_current_nodes*floor(node_growth_rate) new nodes,
and then add number_of_current_nodes*(node_growth_rate%1) in expectation.

edge_growth_rate:
As above, but for edges

algorithm:
Optional replication with one of several alternative generators (not MUSKETEER), that are implemented in algernatives.py
For example, to use the Expected Degrees Model, use:
-p "{'algorithm':alternatives.expected_degree_replicate}"


'algorithm':algorithms.musketeer_snapshots:
Parameter causes editing sequentially: G->R1->R2->...->Rn and return a sequence of snapshots.
The edit rate parameter is applied at every generation step.
Use
'num_snapshots':num_snapshots    (any positive integer)
The output returns Rn.  The attribute .snapshots contains a list of all the graphs from G to Rn.


Algorithm Tuning
----------------
accept_chance_edges:
Chance edges are those edges that inserted regardless of the distance between u and v.
The rate of such edges is determined when computing topological/structural data
but is often spuriously elevated because of computational limitation (see edge horizon)
Graphs like the Watts-Strogatz graph have a few but very significant long-distance edges.
Decreasing accept_chance_edges from 1.0 (default) closer to 0.0 would reduce spurious chance edges.

'algorithm':algorithms.musketeer_iterated_cycles
Alternating editing: calls the main algorithm multiple times passing the output to the input.
The edit and growth rates are split over the cycles 
in order to gradually change the graph, increasing realism.
(e.g. two cycles of [0.05, 0.01] + [0.05, 0.01] instead of one [0.1, 0.02])
Use the specification:
'num_v_cycles':num_v_cycles    (any positive integer)


deep_copying:
When True (default), new coarse nodes (aggregate) precisely match the structure of an existing aggregate,
and this is true at all levels of refinement.
Previously, at each level of refinement, random resampling was applied to interpolate the aggregate.

deferential_detachment_factor:
Default 1.  
Changes the probability of deleting an edge (u,v) to approximately 1/(degree(u) * degree(v) )  
Reducing this closer to 0 would disable deferential detachment mechanism.

dont_cutoff_leafs:
When true, prevents edge editing from removing connection to a leaf node (boolean)
Not recommended, since it skews the degree distribution.

edge_welfare_fraction:
Increases the likelihood that nodes which lost an edge will be given a replacement.  Ranges 0..1
Helps recover from bias in clustering.

enforce_connected:
This parameter (boolean) will insert new edges from small components to the giant component.
The default behavior is to enforce connectedness if the original graph was connected.

An alternative handling of disconnected graphs is to apply musketeer on each of the components of the graph.
Use the parameter 
'algorithm':algorithms.musketeer_on_subgraphs
This keeps their sizes the same.

fine_clustering:
When True (default: False), the addition of edges considers the number of triangles that would be created.
This slightly increases fidelity at somewhat high computational cost.

locality_bias_correction:
Increases the number of triangles generated during edge insertion.  
Ranges -1..1 at each level, similar to edge_edit_rate: [0.6, 0.7]
Positive values help recover from negative bias in clustering.

maintain_edge_attributes:
When True (default False), preserves attributes of edges (weight, strengths, friendship type). 
Attributes of new edges are copied from attributes of existing edges.

maintain_node_attributes
When True (default False), preserves attributes of nodes (such as color, body mass, electrical output etc) 
Attributes of new nodes are copied from attributes of existing nodes.
 
memoriless_interpolation:
When True (default False), interpolation is applied even to nodes and edges which were not edited.
Data on aggregates that was created during coarsening is not used.
This option provides randomization of the network and may be useful for anonymization (no guarantees are given).

minorizing_node_deletion:
When True, deleted nodes are replaced by a tree that connects all their old neighbors.

new_edge_horizon:
This parameter (positive integer) determines the depth of scanning when inserting new edges.
Increasing this value increases fidelity but slows the algorithm, especially when the graph is dense.
The default value is determined automatically based on the density of the graph.

post_processor:
Runs the replica through an algorithm with the pattern f(new_G, original, params) before saving the replica
try 'post_processor':graphutils.color_new_nodes_and_edges

preserve_clustering_on_deletion:
When True (default), significantly helps increase the fidelity of clustering in replicas at a modest overhead and decrease in entropy.
It decreases the probability that an edge (u,v) is deleted when u and v have many mutual neighbors.

retain_intermediates:
When True (default: False), intermediate graphs are retained and accessible from new_G.
  new_G.coarser_graph = the final result of the previous level
  new_G.model_graph       = the input graph to the previous level retain_intermediates (i.e. the graph G_i, which is horizontally on the V cycle)
coarser_graph is a linked list (new_G.coarser_graph.coarser_graph...) which terminates with None at the coarsest level.


stats_report_on_all_levels:
When True (default: False), compares the coarsen graphs at each level, showing how the original and edited versions differ.

suppress_warnings:
When True (default: False), warnings are not displayed and the algorithm is almost completely silent.

verbose:
Enables detailed output (boolean).


Troubleshooting
---------------
Accelerating the computation
* remove the computation of the metrics with "-M False" argument (all metrics);  or in graphutils.py, make some complex metrics optional.
* convert the original graph into a simple format like edgelist
* reduce the new_edge_horizon (reduces fidelity)
* set deferential_detachment_factor close to 0 (reduces fidelity)
* give the authors funding to develop a version in the C language

Clustering is diminished in the replicas
* make sure that the pattern is consistent - there would be and need to be some variance
* reduce the edge editing rate at the finest level: it would improve preservation of all fine-level properties of the graph
* increase locality_bias_correction
* make fine_clustering True
* make preserve_clustering_on_deletion True

Path lengths and related global metrics are diminished in the replicas
* make accept_chance_edges closer to 0.  As a compensation to maintain fidelity, increase new_edge_horizon to 10, 20 or larger.

Support for weighted edges
* currently only used in coarsening
* we are planning to include it in future releases

Support for node and edge attributes
* use the parameters 'maintain_node_attributes':True, 'maintain_edge_attributes':True

Error writing DOT file or pygraphviz error:
* pygraphviz is not currently functional in the Windows platform
* specify an alternative output file such as "-o my_output.elist"
* or, try install/ing pydot package



release_notes.txt
2023-02-20: version 1.3:
------------------------
- Migrated to python3
- added requirements.txt, tests with pytest
- removed lots of extreneous code
- removed redundant community.py and used native NX function
- removed algebraic_distance (temporarily) due to large changes in NX APIs

2013-11-??: Version 1.2:
------------------------
Changes
- parameters algorithm_for_coarsening and algorithm_for_uncoarsening introduce support for user-defined coursening/uncoarsening functions
- additional parameters
- support for control of runningtime bound
- new_alg


2013-07-06: Version 1.11:
------------------------
Changes
- replica.predecessor_graph is renamed to replica.coarser_graph
- computes Jaccard similarity coefficient

2013-03-26: Version 1.1:
------------------------
Changes
-removed edit_method:alternating
-improved reporting of output
-added option: post_processor
-support of deep_copying (not fully tested yet)
-removed obsolete 'error_rate'
-added validation for the names of all parameters given in the input
-cleanup of uncoarsening code
-retain_intermediates parameter now makes the algorithm return intermediate graphs
-disabled deferential detachment by default


2012-11-18: Version 1.02:
------------------------
Changes
-update the help file
-support for growth in the number of edges
-support for editing with a W-cycle (to replace deprecated edit_method:alternating)

Bug fix:
-fixed severe bug when enforcing connected graphs.  affecting primarily edit_method:alternating

2012-10-01: Version 1.01:
------------------------
Changes
-support for Graphviz (.dot) format on systems without pygraphviz, especially Windows
-support for operation on graph components (using algorithms.musketeer_on_components)
-support for generation of snapshots


2012-07-18: Version 1.0:
------------------------
Changes
-support for rendering in Linux (graphviz version 2.26.2)
-improved interpolation: better resampling
-memoriless interpolation for better ananymization

2012-07-17: Version 1.0:
------------------------
Released as 1.0 without changes

2012-06-21: Version 0.92:
---------------------------------
Changes
-more advanced visualization of graphs using commandline option: -v sfdp3d

Bug fixes
-output bug in compare_nets

~sasha

2012-05-14: Version 0.91:
---------------------------------
Changes
-simple support for node and edge attributes through resampling
-new command line options
-accept_chance_edges allows more control

Bug fixes
-

~sasha


2012-04-16: Version 0.90:
---------------------------------
Changes
-New parameter 'preserve_clustering_on_deletion' improves accuracy of the clustering metric in replicas for highly clustered original networks
-Performance improvements by optimization of calls to common methods

Bug fixes
-

~sasha.

2012-04-04: Version 0.89:
---------------------------------
Changes
-Optional algorithm that alternatives between deletion and insertion
-Optional algorithm for fine insertion of edges at distance 2 in order to avoid damaging clustering.
-Option to removing evaluation of network properties (-N)

Bug fixes
-

~sasha.



2012-03-06: Version 0.88:
---------------------------------
Changes
-Deferential detachment and edge_welfare mechanisms.
-New way of computing locality

Bug fixes
-Clustering is handled better

~sasha.


2012-02-13: Version 0.87:
---------------------------------
Changes
- Records the random generator seed in case the replication must be replicated.
- edge weights are now used during coarsening and propagated down the v-cycle
  during coarsening, node u attaches to the neighboring seed with the strong edge weight (seed,u) 
  however, the final graph does not retain the weights
- more straight-forward handling of new edges
- improved debugging support

Bug fixes
- handling of horizon in a special case 

~sasha.


2012-01-29: Version 0.86:
---------------------------------
Changes
- More accurate handling of locality
- Coarsening stops automatically if no deeper edits are given
- Customizable output format

Bug fixes
- 

Known problems
-

~Sasha

2012-01-16: Version 0.85:
---------------------------------
Changes
- 
Bug fixes
- Fixes to locality handling
- Fix to edge and node deflation

Known problems
-

~Sasha


2012-01-14: Version 0.8
---------------------------------
Changes
- Refactoring of main loop
- Disable minorization of deleted nodes
- Revision of the creation of edges

Bug fixes
- Fixes to locality handling

Known problems
- Suspect bug in locality

~Sasha


simpletesters.py
'''
Multiscale Entropic Network Generator 2 (MUSKETEER2)

Copyright (c) 2011-2015 by Alexander Gutfraind and Ilya Safro.
All rights reserved.

Use and redistribution of this file is governed by the license terms in
the LICENSE file found in the project's top-level directory.


Test Scripts

'''

import os
import time
import numpy as np
import numpy.random as npr
import random, sys
import networkx as nx
#import matplotlib
#matplotlib.use('PDF')
#import matplotlib.pylab as pylab
#import pylab
import pdb
import cPickle
import graphutils

np.seterr(all='raise')

timeNow = lambda : time.strftime('%Y_%m_%d__%H_%M_%S', time.localtime())

def integrity_test():
    import algorithms
    print 'Integrity testing ...'
    graphs = {'karate': nx.generators.karate_club_graph(),
              'er200_025': nx.erdos_renyi_graph(n=200, p=0.25, seed=17),
              'er200_0001': nx.erdos_renyi_graph(n=200, p=0.001, seed=42)}

    params = {'verbose':True,
              'node_edit_rate': [0],
              'edge_edit_rate': [0],
              'node_growth_rate': [0],
              'verbose':False}
    for name,G in graphs.items():
        print name
        replica = algorithms.generate_graph(original=G, params=params)

        diff    = graphutils.graph_graph_delta(G, replica)
        assert diff['new_nodes'] == []
        assert diff['del_edges'] == []
        assert diff['new_nodes'] == []
        assert diff['del_edges'] == []

    print 'Integrity test: PASSED'

def iterated_test(seed=None, testparams=None, params=None):
    import algorithms
    print 'Starting iterative replication test...'
    if seed == None:
        seed = npr.randint(1E6)
        print 'Setting random number generator seed: %d'%seed
        random.seed(seed)
        npr.seed(seed)
    if params == None:
        params = {}
    defparams = {'edge_edit_rate':[0.05, 0.05], 'node_edit_rate':[0.05, 0.05], 'verbose':False, }
    defparams.update(params)
    params = defparams
    print 'params'
    print params
    if testparams == None:
        testparams = {}
    if 'G' not in testparams:
        nn = 1000
        p  = 0.01
        G = nx.erdos_renyi_graph(nn, p)
        G.add_edges_from(nx.path_graph(nn).edges())
    else:
        G = testparams['G']
    alg = testparams.get('algorithm', algorithms.generate_graph)

    num_rounds = testparams.get('num_rounds', 10)
    print 'Round: 1. Initial graph ' + getattr(G, 'name', '_')
    for trial in xrange(2, num_rounds+1):
        new_G = alg(original=G, params=params)
        seed = npr.randint(1E6)
        print 'Round: %d. New seed: %d'%(trial,seed)
        random.seed(seed)
        npr.seed(seed)
    print 'PASSED'

def smoke_test():
    import algorithms
    print 'Smoke testing ...'
    graphs = {'karate': nx.generators.karate_club_graph(),
              'er200_025': nx.erdos_renyi_graph(n=200, p=0.25, seed=42),
              'er200_0001': nx.erdos_renyi_graph(n=200, p=0.001, seed=42)}

    params = {'verbose':False,
              'node_edit_rate': [0.1/(1.+i) for i in xrange(100)],
              'edge_edit_rate': [0.1/(1.+i) for i in xrange(100)],
              'node_growth_rate': [0.1/(1.+i) for i in xrange(100)]}
    for name,G in graphs.items():
        print name
        #print '  nn=%d,ne=%d'%(G.number_of_nodes(), G.number_of_edges())
        replica = algorithms.generate_graph(original=G, params=params)
        #print '  nn=%d,ne=%d'%(replica.number_of_nodes(), replica.number_of_edges())
        assert list(G.selfloop_edges()) == []

    assert 0 == os.system(graphutils.MUSKETEER_EXAMPLE_CMD)


    print 'Smoke test: PASSED'
    print
    return

#we use a dict, for possible future detailed testing code, such as testing for numerical range
valid_params = {'accept_chance_edges':None,
                'algorithm':None,
                'algorithm_for_coarsening':None, #TODO: document
                'algorithm_for_uncoarsening':None, #TODO: document
                'coarsening_density_limit':None, #TODO: document
                'component_is_edited':None, #TODO: document
                'deep_copying':None,
                'deferential_detachment_factor':None,
                'do_coarsen_tester':None, #TODO: document
                'do_uncoarsen_tester':None, #TODO: document
                'dont_cutoff_leafs':None,
                'edge_edit_rate':None,
                'edge_growth_rate':None,
                'edge_welfare_fraction':None,
                'edit_edges_tester':None, #TODO: document
                'edit_nodes_tester':None, #TODO: document
                'enforce_connected':None,
                'fine_clustering':None,
                'locality_algorithm':None,
                'locality_bias_correction':None,
                'long_bridging':None, #TODO: document
                'maintain_edge_attributes':None,
                'maintain_node_attributes':None,
                'matching_algorithm':None, #TODO document
                'memoriless_interpolation':None,
                'metric_runningtime_bound':None, #TODO: document
                'minorizing_node_deletion':None, #TODO: document
                'new_edge_horizon':None,
                'node_edit_rate':None,
                'node_growth_rate':None,
                'num_deletion_trials':None, #TODO: document
                'num_insertion_trials':None, #TODO: document
                'num_insertion_searches_per_distance':None, #TODO: document
                'num_pairs_to_sample':None, #TODO: document
                'num_snapshots':None,
                'num_trial_particles':None, #TODO: document
                'num_v_cycles':None,
                'post_processor':None,
                'preserve_clustering_on_deletion':None,
                'reporting_precision':None, #TODO: document
                'revise_graph_tester':None, #TODO: document
                'retain_intermediates':None,
                'seed_threshold_1':None, #TODO: document
                'seed_threshold_2':None, #TODO: document
                'search_method':None, #TODO: document
                'skip_param_sanity_check':False, #TODO: document
                'stats_report_on_all_levels':None, #presumes retain_intermediates
                'triangle_distribution_limit':None, #TODO: document
                'suppress_warnings':None,
                'verbose':None,
                'weighted_step':None,
                }

def validate_params(params):
    bad_params = False
    for k in params:
        if k not in valid_params and (not str(k).startswith('_')):
            if params.get('verbose', True):
                print 'Unknown or undocumented parameter: %s'%k
                print 'Hint: for a list of valid parameters, see valid_params variable in simpletesters.py'
                print
            bad_params = True

    if params.get('memoriless_interpolation', False) and params.get('deep_copying', True):
        if params.get('verbose', True):
            print 'Memoriless_interpolation=True requires deep_copying=False'
        bad_params = True

    if params.get('stats_report_on_all_levels', False) and not params.get('retain_intermediates'):
        if params.get('verbose', True):
            print 'Making retain_intermediates=True'
        params['retain_intermediates'] = True


    if bad_params and not params.get('skip_param_sanity_check', False):
        if params.get('verbose', True):
            print 'Existing ...'
        sys.exit(1)


if __name__ == '__main__':
    smoke_test()
    integrity_test()
    iterated_test()
